### PR TITLE
Port `all` kernel to structured kernels.

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59938 Port `argmin` kernel to structured kernels.
* #59937 Port `argmax` to structured kernels.
* #59373 Using meta checks for unary `torch.all` and `torch.any`.
* #59372 Port `any` kernel to structured kernels.
* **#59371 Port `all` kernel to structured kernels.**

Tracking issue: #55070

Differential Revision: [D29104399](https://our.internmc.facebook.com/intern/diff/D29104399)

### CODE REVIEW

File: aten/src/ATen/native/ReduceOpsUtils.h [hunk 1]
```diff
@@ -301,5 +301,72 @@ static void zero_numel_tensor_resize(Tensor& result, Tensor& result_indices,
...
+        " and ",
+        toString(out_dtype),
+        ".");
+  }
+  // dim={} performs an all-reduce, same as dim=None
+  IntArrayRef dim = dim_opt.value_or(IntArrayRef{});
+  int64_t ndim = self.dim();
+  auto mask = at::native::make_dim_mask(dim, ndim);
+  auto shape = at::native::shape_from_dim_mask(self, mask, keepdim);
+  meta.set_output(shape, self.options().dtype(out_dtype));
```
Comment (bdhirsh) [inline aten/src/ATen/native/ReduceOpsUtils.h:332]: why not have `all` inherit from TensorIteratorBase?

To make that work, you're going to need to update `TensorIterator.h/cpp` and write separate functions for `TensorIteratorBase::reduce_op`. You can take a look at how we split out similar functionality by looking at, e.g., `TensorIterator::binary_op` and `TensorIteratorBase::build_binary_op`.

File: aten/src/ATen/native/ReduceOpsUtils.h [hunk 2]
```diff
@@ -301,5 +301,72 @@ static void zero_numel_tensor_resize(Tensor& result, Tensor& result_indices,
   at::native::resize_output(result, sizes);
   at::native::resize_output(result_indices, sizes);
 }
+} // native
 
-}}  // at::native
+namespace meta {
+
+static void check_reduction_shape(
+    impl::MetaBase& meta,
```
Comment (bdhirsh) [inline aten/src/ATen/native/ReduceOpsUtils.h:309]: nit: I think it's cleaner to leave the meta object out of the helper. So the helper can just make various assertions (and return data if it needs to), and you can call `set_output()` or `build_reduce_op` directly inside of `TORCH_META_FUNC`

File: aten/src/ATen/native/ReduceOps.cpp [hunk 3]
```diff
@@ -1148,44 +1194,12 @@ Tensor all(const Tensor& self) {
...
-      auto iter = make_reduction(
-          "all", result, self, dim, keepdim, self.scalar_type(), out_dtype);
-      return _all(result, iter);
-    }
-    auto iter =
-        make_reduction("all", result, self, dim, keepdim, /*out_dtype=*/out_dtype);
-    return _all(result, iter);
+TORCH_IMPL_FUNC(all_out)
+(const Tensor& self, int64_t dim, bool keepdim, const Tensor& result) {
+  auto iter = get_reduction_iter(self, result, dim, keepdim);
```
Comment (bdhirsh) [inline aten/src/ATen/native/ReduceOps.cpp:1199]: If you can make `all` subclass from TensorIteratorBase, then you won't need this (and you won't risk duplicating the code between the meta and impl methods.

**Comment by ezyang** [comment]
Yes, echoing Brian, this should be doing the TensorIterator path in https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md#handling-tensoriterator

File: aten/src/ATen/native/ReduceOpsUtils.h (hunk 1 above)
Comment (ysiraichi) [reply aten/src/ATen/native/ReduceOpsUtils.h:332]: Actually, I did try that at first, but failed.
The problem is that `all` is resized outside the `TensorIteratorBase` (and it expects this resizing to happen).
https://github.com/pytorch/pytorch/blob/0d7d316dc1a889937227d29978ee86315f33bc24/aten/src/ATen/TensorIterator.cpp#L178

So, I tried to call `set_output` before calling `build_reduce_op`. However, I was getting the following error when calling `set_output`:

```python
RuntimeError: op.original_tensor->is_same(t)INTERNAL ASSERT FAILED at "../aten/src/ATen/TensorIterator.cpp":1391, please report a bug to PyTorch.
```

Which, I think, is due to `TensorIteratorBase::operands_` not being initialized. When using `gdb`, I checked, for example, the value of `op.will_resize` (`OperandInfo`) which evaluated to 180 (yes, even though it's a `bool`).

In summary, as far as I understood, calling `set_output` directly calls `structured_all_out_functional::set_output`, which calls `structured_all_out::set_output`, which fallbacks to `TensorIteratorBase::set_output`. The latter's `operands_` is not initialized, therefore I get that error.

Am I missing anything here?

Comment (bdhirsh) [reply aten/src/ATen/native/ReduceOpsUtils.h:332]: oof okay, thanks for pointing that out. I see `all` doing two things with TensorIterator that look unusual:

(1) it [resizes the output tensor](https://github.com/pytorch/pytorch/blob/f2406240800e7cb3b806a4de94e9ef3d4838c547/aten/src/ATen/native/ReduceOpsUtils.h#L197) before creating the TensorIterator, like you said. (It also makes `TensorIteratorBase::allocate_or_resize_outputs()` a no-op, which is a little confusing).

(2) It [creates a view off of the resized output tensor](https://github.com/pytorch/pytorch/blob/f2406240800e7cb3b806a4de94e9ef3d4838c547/aten/src/ATen/native/ReduceOpsUtils.h#L198) using `review_reduce_result()`, and passes the view into TensorIterator.

If we want structured reduction ops to use TensorIterator, I think these are both issues.

(1) is an issue because our only structured kernel interface to directly resize the output tensor is with `set_output()`, but that in turn calls `TensorIteratorBase::set_output()`, which is problematic because we're calling it before constructing the actual TensorIterator. That's why you got the internal assert failure you pointed out above: `op.original_tensor` hasn't been properly set yet because we haven't called `reduce_op()` yet.

(2) Is also a problem, because TensorIteratorBase currently expects that the output tensor that you pass into it is the same as the underlying tensor in `outputs_[0]` (I tested it out locally and confirmed that we get an assert failure). That assertion lives [here](https://github.com/pytorch/pytorch/blob/4f79270b8933a1e580343917137f0131754cda21/aten/src/ATen/TensorIterator.cpp#L1392)

So I think there are two options:



**Option 1:** Keep stuff as is.

This feels of sub-optimal but it's also the current state. Continue to do the resizing/restriding directly in these helper functions outside of TensorIterator, and have this reduction op (and future ones?) inherit from `MetaBase`, not `TensorIteratorBase`. One downside is that we’ll have structured ops that continue to use TensorIterator (and not TensorIteratorBase).

It looks like CI is passing even though you don't directly use TensorIterator anywhere in the meta function. I was kind of surprised, since I would have thought that the meta API would miss out on some restriding/type promotion edge cases that TI normally handles, but from looking at `TensorIterator::reduce_op()`, it might just be that the manual checks that you're doing in `check_reduction_shape` handle everything. `TensorIteratorBase::allocate_or_resize_outputs()` also looks like a no-op, since you already manually resized before calling into it.



**Option 2:** Get TensorIterator to fully handle reduction op meta logic.
It seems like we'd need to make some pretty intrusive changes to TensorIteratorBase if we wanted it to be able to fully handle structured reduction ops. Specifically, right now TI relies on some preconditions (output resizing and restriding) to happen before constructing the TensorIterator object. We would need TensorIteratorBase to actually do both of those things for reduction ops.

I'm kind of thinking it would be worth doing the refactor at some point, but not necessary for this PR.

@ezyang  @ngimel what do you think?

Comment (ezyang) [reply aten/src/ATen/native/ReduceOpsUtils.h]: On further reflection, I think I might be pretty OK with the current implementation. More shortly.

Comment (ezyang) [reply aten/src/ATen/native/ReduceOpsUtils.h]: So, here is why I think it's OK not to inherit from TensorIteratorBase: complexity is conserved. @bdhirsh, you already observed this in: "It looks like CI is passing even though you don't directly use TensorIterator anywhere in the meta function. I was kind of surprised". `TORCH_META_FUNC2` is pretty short, it doesn't have most of the fancy logic that TensorIterator normally handles, because prior to this PR, we were already shortcircuiting most of the logic in TI. Yes there's monkey business with `make_reduction` and `make_reduction_from_out_ty`, but that's also preexisting monkey business from the old PR.

> One downside is that we’ll have structured ops that continue to use TensorIterator (and not TensorIteratorBase).

I agree this is a downside, although it isn't going to very pressing until we get close to finishing porting all TensorIterator off; in particular, there are cases that allocate TI on the fly that we'd have to figure out what to do with. Once those are all handled, then we can talk more about how to kill stock TI entirely here as well.

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29104399).

**Review by ezyang — APPROVED** [review]
